### DISCUSSION TITLE
One-liner with ```which``` behaves differently from just running in bash

### DISCUSSION THREAD

**Discussion by artsiommiksiuk** [casey/just#2184] [Q&A]
I have this recipe:

```justfile
PYTHON_VER := "3.10"
test:
    which python{{PYTHON_VER}}-config
    mkdir -p .venv/bin && ln -sf $(which python{{PYTHON_VER}}-config) .venv/bin/python{{PYTHON_VER}}-config
    which python{{PYTHON_VER}}-config
```

Which outputs on the first run

```bash
which python3.10-config
/opt/homebrew/bin/python3.10-config
mkdir -p .venv/bin && ln -sf $(which python3.10-config) .venv/bin/python3.10-config
which python3.10-config
/Users/app/.venv/bin/python3.10-config
```

and on the second immediate run
```bash
which python3.10-config
/Users/app/.venv/bin/python3.10-config
mkdir -p .venv/bin && ln -sf $(which python3.10-config) .venv/bin/python3.10-config
which python3.10-config
/opt/homebrew/bin/python3.10-config
```

> Notice ```which``` output is now reversed

while running the same hydrated command in bash behaves normal - no reversing and which is always resolved to .venv folder. Why does it happen?

Am I confusing something?

**Comment by wellkilo** [comment]
Each non-shebang recipe line is executed in a **fresh shell**. That explains why the three which calls do not share shell state, but the alternating result specifically points to your shell initialization/environment changing PATH between those fresh invocations (for example a virtual-environment hook such as direnv/pyenv/mise, or a shell startup file that prepends/removes .venv/bin).

Use one shell process for the whole recipe and resolve the executable before creating the link:

    test:
      #!/usr/bin/env bash
      set -euo pipefail
      target=$(command -v python{{PYTHON_VER}}-config)
      printf '%s\n' "$target"
      mkdir -p .venv/bin
      ln -sf "$target" .venv/bin/python{{PYTHON_VER}}-config
      command -v python{{PYTHON_VER}}-config

There is another trap in the original: once .venv/bin/python3.10-config is first on PATH, which may resolve the destination symlink itself. Re-linking the file to that value can create a self-referential symlink. Capturing the intended external target first—or using the known absolute path—avoids that.

To diagnose the alternating PATH, temporarily print printf '%s\n' "$PATH" on each original recipe line. You will likely see that the startup hook treats the new .venv state differently on each shell invocation.

Just documents the execution model explicitly: [each line is executed by a fresh shell](https://github.com/casey/just/blob/5969e396f4e5c98b64a5df929e3ee67629295d6e/README.md#L3067-L3077), while a shebang recipe runs the whole body as one script.
